jellyfin-api/src/commonMain/kotlin/org/jellyfin/sdk/api/sockets/SocketInstance.kt

```diff
@@ -0,0 +1,243 @@
+
+	private suspend fun updateConnectionState(): Boolean {
+		// Don't update when an error occurred, client should manually call [reconnect]
+		if (state.value == SocketInstanceState.ERROR) {
+			logger.info { "Unable to update connection state: state is error" }
+			return false
+		}
+
+		// Remove listeners that don't want credential changes
+		if (credentialsChanged) listenerHelper.reportCredentialChangedReconnect()
+
+		val connected = connection != null // TODO make smarter (handle disconnect etc.)
+
+		when {
+			// Stop if there's no listeners
+			listenerHelper.listeners.isEmpty() -> connection?.disconnect()
+			// Reconnect when credentials changed or not connected
+			credentialsChanged || !connected -> reconnect()
+			// Update subscriptions when not reconnecting or disconnecting
+			else -> updateSubscriptions()
+		}
+
+		// Make sure credentials changed is set to false
+		credentialsChanged = false
+
+		return true
+	}
+
+	private suspend fun updateSubscriptions() {
+		val subscriptions = listenerHelper.subscriptions
+
+		// Remove subscriptions not in use anymore
+		for (subscription in listenerHelper.activeSubscriptions.reversed()) {
+			if (!subscriptions.contains(subscription)) {
+				logger.info { "Removing subscription for ${subscription.messageType.simpleName}" }
+				publish(subscription.createStopMessage())
+				listenerHelper.activeSubscriptions.remove(subscription)
+			}
+		}
+
+		// Add new subscriptions
+		// Period is not configurable as it is barely used by the server
+		// The initialDelay is never used in fact
+		val period = PeriodicListenerPeriod(0, MESSAGE_INTERVAL)
+		for (subscription in subscriptions) {
+			if (!listenerHelper.activeSubscriptions.contains(subscription)) {
+				logger.info { "Adding subscription for ${subscription.messageType.simpleName}" }
+				publish(subscription.createStartMessage(period))
+				listenerHelper.activeSubscriptions.add(subscription)
+			}
+		}
+	}
+
+	public suspend fun reconnect() {
+		logger.debug { "Reconnect requested" }
+
+		// Already connecting
+		if (state.value == SocketInstanceState.CONNECTING) return
+
+		// Explicitly stopped
+		if (state.value == SocketInstanceState.STOPPED) throw SocketStoppedException()
+
+		logger.info { "Reconnecting" }
+		_state.value = SocketInstanceState.CONNECTING
+
+		connection?.disconnect()
+
+		connection = socketConnectionFactory.create(
+			api.httpClientOptions,
+			incomingMessages,
+			outgoingMessages,
+			coroutineContext,
+		).apply {
+			val connected = connect(UrlBuilder.buildUrl(
+				baseUrl = requireNotNull(baseUrl),
+				pathTemplate = SOCKET_URL,
+				queryParameters = mapOf(
+					QUERY_DEVICE_ID to deviceInfo.id,
+					QUERY_ACCESS_TOKEN to accessToken,
+				)
+			).replace(Regex("^http"), "ws"))
+
+			if (connected) {
+				messageForwardJob?.cancel()
+				messageForwardJob = coroutineScope.launch {
+					for (message in incomingMessages) forwardMessage(message)
+				}
+				listenerHelper.activeSubscriptions.clear()
```

Comment:
It does this on connect to make sure the state is correct when initializing a connection. it could be cleared on disconnect too but it doesn't make that much of a difference.

Reply:
Ok then